<a href="https://colab.research.google.com/github/xpertdesh/ml-class21/blob/main/labs/pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Intro to Pandas
![](https://raw.githubusercontent.com/zacharski/ml-class/master/labs/pics/Pandas_logo.svg)

(logo by [Marc Garcia](https://github.com/pandas-dev/pandas/blob/master/web/pandas/static/img/pandas.svg))

The Pandas Library is built on top of Numpy and is designed to make working with data fast and easy. Like Numpy, the library includes data structures and functions to manipulate that data.

As we learned in the Numpy Notebook, we need to load in the library before we can use it.

In [ ]:
from pandas import Series, DataFrame
import numpy as np
import pandas as pd

Let's dissect the code above.

The `Series` and `DataFrame` datatypes are commonly used so we import them directly with

```
from pandas import Series, DataFrame
```

For all other datatypes and functions in the library the `pd` prefix is commonly used so we import that with

```
import pandas as pd
```

## Series
A series is a 1d array-like object

Let's consider the heights (in cm) of Japan's Women's Basketball Team at the 2016 Olympics.      

We can create a series in a number of ways.

##### directly from a Python list

In [ ]:
japan = [173, 182, 185, 176, 183, 165, 191, 177, 165, 161, 175, 189]
athletesHeight = Series(japan)

We could also have done

In [ ]:
athletesHeight = Series([173, 182, 185, 176, 183, 165, 191, 177, 165, 161, 175, 189])

In either case we can see the value of the Series athletesHeight

In [ ]:
 athletesHeight

0     173
1     182
2     185
3     176
4     183
5     165
6     191
7     177
8     165
9     161
10    175
11    189
dtype: int64

As like arrays you are familiar with, the left number is the index and the right the value. And we can find the value at a particular index by the usual:

In [ ]:
athletesHeight[3]

176

#### specifying indices
Instead of the index 0, 1, 2, 3 ... you can specify your own index values. For example, we can label them 'Aiko'. 'Ayumi', "Fujiko" ... etc. 

In [ ]:
athletes2 = Series(japan, index = ['Aiko', 'Ayumi', 'Fujiko', 'Hiroko', 'Itsumi', 'Junko', 'Kanae','Kotori', 'Mieko', 'Momoko', 'Nao', 'Rei'])
athletes2

Aiko      173
Ayumi     182
Fujiko    185
Hiroko    176
Itsumi    183
Junko     165
Kanae     191
Kotori    177
Mieko     165
Momoko    161
Nao       175
Rei       189
dtype: int64

The names we see are not another column of the data. We can see the shape of athletes2 by:

In [ ]:
athletes2.shape

(12,)

This shows that athletes2 is a one dimensional matrix and that dimension has a length of 12. So the names we see are not values in a column but rather the indices.


Let's use the index to get the Height of Junko:

In [ ]:
athletes2['Junko']

165

How would you get the height of Nao?



## DataFrame
DataFrames are the most important data structure of Pandas and are simply
a table or spreadsheet like structure.  A DataFrame represents a table like:

make | mpg | cylinders | HP | 0-60 
:---- | :---: | :---: | :---: | :---: 
Fiat | 38 | 4 | 157   | 6.9 
Ford F150 | 19 | 6 | 386 | 6.3 
Mazda 3 | 37 | 4 | 155 |  7.5 
Ford Escape | 27 | 4 | 245 | 7.1 
Kia Soul | 31 | 4 | 164 | 8.5 

A common way to create a DataFrame is to use a python dictionary as follows:

In [ ]:
cars = {'make': ['Fiat 500', 'Ford F-150', 'Mazda 3', 'Ford Escape', 'Kia Soul'],
        'mpg': [38, 19, 37, 27, 31],
        'cylinders': [4, 6, 4, 4, 4],
        'HP': [157, 386, 155, 245, 164],
        '0-60': [6.9, 6.3, 7.5, 7.1, 8.5]}


Just to check that you are paying attention, what is the code to show the MPG of a Ford F-150? (so `cars['make'][3]` will print out something, what will print out the MPG of a Ford F-150?

19

and now we will create a DataFrame from the `cars` Python dictionary:

In [ ]:
df = DataFrame(cars)
df


make  mpg  cylinders   HP  0-60
0     Fiat 500   38          4  157   6.9
1   Ford F-150   19          6  386   6.3
2      Mazda 3   37          4  155   7.5
3  Ford Escape   27          4  245   7.1
4     Kia Soul   31          4  164   8.5

Prior to my life with Pandas, I would represent a table like the above one as:

In [ ]:
prePandas = [{'make': 'Fiat', 'mpg': 38, 'cylinders': 4, 'HP': 157, '0-60': 6.9},
            {'make': 'Ford F-150', 'mpg': 19, 'cylinders': 6, 'HP': 386, '0-60': 6.3},
            {'make': 'Mazda 3', 'mpg': 37, 'cylinders': 4, 'HP': 155, '0-60': 7.5},
            {'make': 'Ford Escape', 'mpg': 27, 'cylinders': 4, 'HP': 245, '0-60': 7.1},
            {'make': 'Kia Soul', 'mpg': 31, 'cylinders': 4, 'HP': 164, '0-60':8.5}]

prePandas[0]['make']

'Fiat'

In the prePandas scheme the data is organized first by rows. That seemed logical to me since each row represents an object and is how we organize data in an SQL database. In the Pandas representation the data is organized by columns. 

Perhaps you noticed that the columns are displayed in alphabetically sorted order. If we want to specify a column order we can do so:




In [ ]:
df2 = DataFrame(cars, columns=['make', 'mpg', 'cylinders', 'HP', '0-60' ])
df2

make  mpg  cylinders   HP  0-60
0     Fiat 500   38          4  157   6.9
1   Ford F-150   19          6  386   6.3
2      Mazda 3   37          4  155   7.5
3  Ford Escape   27          4  245   7.1
4     Kia Soul   31          4  164   8.5

We can also create a DataFrame from a set of Series objects:

In [ ]:
cars2 = {'make': Series(['Fiat 500', 'Ford F-150', 'Mazda 3', 'Ford Escape', 'Kia Soul']),
        'mpg': Series([38, 19, 37, 27, 31]),
        'cylinders': Series([4, 6, 4, 4, 4]),
        'HP': [157, 386, 155, 245, 164],
        '0-60': Series([6.9, 6.3, 7.5, 7.1, 8.5])}
        
df3 = DataFrame(cars2)


<h3 style="color:red">Q1. You Try</h3>
<span style="color:red">Make a DataFrame representing the data in the following table:</span>

Athlete | Sport | Height | Weight 
----: | :---: | :---: | :---:
Aly Raisman |	Gymnastics | 	62 | 115
Crystal Langhorne |	Basketball |	74 |	190
Diana Taurasi	| Basketball	| 72	| 163
Paula Radcliffe	| Track	| 68 |	120
Simone Biles | Gymnastics | 57 | 104
Madison Kocian | Gymnastics |62 | 101
Amy Cragg |Track | 64 | 99


## reading data from different sources.
### csv file:
We can create a DataFrame from a CSV file (comma separated values), which is a common format for datasets. 


We use the `pd.read_csv` function to do so.  `pd.read_csv` can read a csv file from either your local machine or the web. Let's start with the web.

#### Reading a CSV file from the web.
To read a file from the web, we simply provide a URL:


In [ ]:
df4 = pd.read_csv('bhttps://raw.githubusercontent.com/zacharski/ml-class/master/data/athletes.csv')
df4

Name       Sport  Height  Weight
0      Asuka Teramoto  Gymnastics      54      66
1    Brittainey Raven  Basketball      72     162
2            Chen Nan  Basketball      78     204
3       Gabby Douglas  Gymnastics      49      90
4    Helalia Johannes       Track      65      99
5     Irina Miketenko       Track      63     106
6       Jennifer Lacy  Basketball      75     175
7        Kara Goucher       Track      67     123
8         Linlin Deng  Gymnastics      54      68
9       Nakia Sanford  Basketball      76     200
10         Nikki Blue  Basketball      68     163
11    Qiushuang Huang  Gymnastics      61      95
12     Rebecca Tunney  Gymnastics      58      77
13        Rene Kalmer       Track      70     108
14    Shanna Crossley  Basketball      70     155
15   Shavonte Zellous  Basketball      70     155
16    Tatyana Petrova       Track      63     108
17        Tiki Gelana       Track      65     106
18    Valeria Straneo       Track      66      97
19    Viktoria Komova  Gymnastics      61      76
20          Amy Cragg       Track      64      99
21       Simone Biles  Gymnastics      57     104
22     Madison Kocian  Gymnastics      62     101
23  Elena Delle Donne  Basketball      77     188
24   Seimone Augustus  Basketball      72     166
25     Desiree Linden       Track      61      97
26   Shalane Flanagan       Track      65     106
27   Laurie Hernandez  Gymnastics      60     106

Sometimes the csv file has a header row as was the case in the example above. That file starts

    Name,Sport,Height,Weight
    Asuka Teramoto,Gymnastics,54,66
    Brittainey Raven,Basketball,72,162
    Chen Nan,Basketball,78,204
    Gabby Douglas,Gymnastics,49,90
    



In [ ]:
!curl https://raw.githubusercontent.com/zacharski/ml-class/master/data/athletes.csv

Name,Sport,Height,Weight
Asuka Teramoto,Gymnastics,54,66
Brittainey Raven,Basketball,72,162
Chen Nan,Basketball,78,204
Gabby Douglas,Gymnastics,49,90
Helalia Johannes,Track,65,99
Irina Miketenko,Track,63,106
Jennifer Lacy,Basketball,75,175
Kara Goucher,Track,67,123
Linlin Deng,Gymnastics,54,68
Nakia Sanford,Basketball,76,200
Nikki Blue,Basketball,68,163
Qiushuang Huang,Gymnastics,61,95
Rebecca Tunney,Gymnastics,58,77
Rene Kalmer,Track,70,108
Shanna Crossley,Basketball,70,155
Shavonte Zellous,Basketball,70,155
Tatyana Petrova,Track,63,108
Tiki Gelana,Track,65,106
Valeria Straneo,Track,66,97
Viktoria Komova,Gymnastics,61,76
Amy Cragg,Track,64,99
Simone Biles,Gymnastics,57,104
Madison Kocian,Gymnastics,62,101
Elena Delle Donne,Basketball,77,188
Seimone Augustus,Basketball,72,166
Desiree Linden,Track,61,97
Shalane Flanagan,Track,65,106
Laurie Hernandez,Gymnastics,60,106

As you can see we can preface any Unix command with a bang (!) to have it execute in our Notebook.  This is amazingly handy:



In [ ]:
!ls

In [ ]:
!pwd

Back to Pandas.

Sometimes the csv file does not have a header row. So for example, a csv file might have the contents:

Asuka Teramoto,Gymnastics,54,66
Brittainey Raven,Basketball,72,162
Chen Nan,Basketball,78,204
Gabby Douglas,Gymnastics,49,90
In that case you specify the names of the columns using the names attribute:

In [ ]:
athletes = pd.read_csv('https://raw.githubusercontent.com/zacharski/ml-class/master/data/athletesNoHeader.csv', names=['Name', 'Sport', 'Height', 'Weight'])
athletes

Name       Sport  Height  Weight
0      Asuka Teramoto  Gymnastics      54      66
1    Brittainey Raven  Basketball      72     162
2            Chen Nan  Basketball      78     204
3       Gabby Douglas  Gymnastics      49      90
4    Helalia Johannes       Track      65      99
5     Irina Miketenko       Track      63     106
6       Jennifer Lacy  Basketball      75     175
7        Kara Goucher       Track      67     123
8         Linlin Deng  Gymnastics      54      68
9       Nakia Sanford  Basketball      76     200
10         Nikki Blue  Basketball      68     163
11    Qiushuang Huang  Gymnastics      61      95
12     Rebecca Tunney  Gymnastics      58      77
13        Rene Kalmer       Track      70     108
14    Shanna Crossley  Basketball      70     155
15   Shavonte Zellous  Basketball      70     155
16    Tatyana Petrova       Track      63     108
17        Tiki Gelana       Track      65     106
18    Valeria Straneo       Track      66      97
19    Viktoria Komova  Gymnastics      61      76
20          Amy Cragg       Track      64      99
21       Simone Biles  Gymnastics      57     104
22     Madison Kocian  Gymnastics      62     101
23  Elena Delle Donne  Basketball      77     188
24   Seimone Augustus  Basketball      72     166
25     Desiree Linden       Track      61      97
26   Shalane Flanagan       Track      65     106
27   Laurie Hernandez  Gymnastics      60     106

### Reading a CSV file the local machine
First, let's get that file onto our local machine:

In [ ]:
curl https://raw.githubusercontent.com/zacharski/ml-class/master/data/athletes.csv > localAthletes.csv

SyntaxError: invalid syntax (<ipython-input-17-e03c0b89ef56>, line 1)

Hmm. That didn't work. Can you fix the error and rerun that cell?

---
Now we can specify the local file using `pd=read_csv`

In [ ]:
d6 = pd.read_csv('localAthletes.csv')
d6

Name       Sport  Height  Weight
0      Asuka Teramoto  Gymnastics      54      66
1    Brittainey Raven  Basketball      72     162
2            Chen Nan  Basketball      78     204
3       Gabby Douglas  Gymnastics      49      90
4    Helalia Johannes       Track      65      99
5     Irina Miketenko       Track      63     106
6       Jennifer Lacy  Basketball      75     175
7        Kara Goucher       Track      67     123
8         Linlin Deng  Gymnastics      54      68
9       Nakia Sanford  Basketball      76     200
10         Nikki Blue  Basketball      68     163
11    Qiushuang Huang  Gymnastics      61      95
12     Rebecca Tunney  Gymnastics      58      77
13        Rene Kalmer       Track      70     108
14    Shanna Crossley  Basketball      70     155
15   Shavonte Zellous  Basketball      70     155
16    Tatyana Petrova       Track      63     108
17        Tiki Gelana       Track      65     106
18    Valeria Straneo       Track      66      97
19    Viktoria Komova  Gymnastics      61      76
20          Amy Cragg       Track      64      99
21       Simone Biles  Gymnastics      57     104
22     Madison Kocian  Gymnastics      62     101
23  Elena Delle Donne  Basketball      77     188
24   Seimone Augustus  Basketball      72     166
25     Desiree Linden       Track      61      97
26   Shalane Flanagan       Track      65     106
27   Laurie Hernandez  Gymnastics      60     106

Suppose we want that file in a data directory. Let's go ahead and create the directory and move the file there.

In [ ]:
!mkdir data
!mv localAthletes.csv data

Now when we load the file we need to give more of a path:

In [ ]:
d6 = pd.read_csv('data/localAthletes.csv')
d6

Name       Sport  Height  Weight
0      Asuka Teramoto  Gymnastics      54      66
1    Brittainey Raven  Basketball      72     162
2            Chen Nan  Basketball      78     204
3       Gabby Douglas  Gymnastics      49      90
4    Helalia Johannes       Track      65      99
5     Irina Miketenko       Track      63     106
6       Jennifer Lacy  Basketball      75     175
7        Kara Goucher       Track      67     123
8         Linlin Deng  Gymnastics      54      68
9       Nakia Sanford  Basketball      76     200
10         Nikki Blue  Basketball      68     163
11    Qiushuang Huang  Gymnastics      61      95
12     Rebecca Tunney  Gymnastics      58      77
13        Rene Kalmer       Track      70     108
14    Shanna Crossley  Basketball      70     155
15   Shavonte Zellous  Basketball      70     155
16    Tatyana Petrova       Track      63     108
17        Tiki Gelana       Track      65     106
18    Valeria Straneo       Track      66      97
19    Viktoria Komova  Gymnastics      61      76
20          Amy Cragg       Track      64      99
21       Simone Biles  Gymnastics      57     104
22     Madison Kocian  Gymnastics      62     101
23  Elena Delle Donne  Basketball      77     188
24   Seimone Augustus  Basketball      72     166
25     Desiree Linden       Track      61      97
26   Shalane Flanagan       Track      65     106
27   Laurie Hernandez  Gymnastics      60     106

## Missing Data

In real machine learning tasks, we often encounter missing values.

For example, suppose we didn't know Brittainey Raven's height. In that case our CSV file would start

    Name,Sport,Height,Weight
    Asuka Teramoto,Gymnastics,54,66
    Brittainey Raven,Basketball,,162
    Chen Nan,Basketball,78,204

with the double comma on the Brittainey line representing the missing data. When we read that file.



In [ ]:
df7 = pd.read_csv('https://raw.githubusercontent.com/zacharski/ml-class/master/data/athletesMissingValue.csv')
df7

Name       Sport  Height  Weight
0      Asuka Teramoto  Gymnastics    54.0      66
1    Brittainey Raven  Basketball     NaN     162
2            Chen Nan  Basketball    78.0     204
3       Gabby Douglas  Gymnastics    49.0      90
4    Helalia Johannes       Track    65.0      99
5     Irina Miketenko       Track     NaN     106
6       Jennifer Lacy  Basketball    75.0     175
7        Kara Goucher       Track    67.0     123
8         Linlin Deng  Gymnastics    54.0      68
9       Nakia Sanford  Basketball    76.0     200
10         Nikki Blue  Basketball    68.0     163
11    Qiushuang Huang  Gymnastics    61.0      95
12     Rebecca Tunney  Gymnastics     NaN      77
13        Rene Kalmer       Track    70.0     108
14    Shanna Crossley  Basketball    70.0     155
15   Shavonte Zellous  Basketball    70.0     155
16    Tatyana Petrova       Track    63.0     108
17        Tiki Gelana       Track    65.0     106
18    Valeria Straneo       Track    66.0      97
19    Viktoria Komova  Gymnastics    61.0      76
20          Amy Cragg         NaN    64.0      99
21       Simone Biles  Gymnastics    57.0     104
22     Madison Kocian  Gymnastics    62.0     101
23  Elena Delle Donne  Basketball    77.0     188
24   Seimone Augustus  Basketball    72.0     166
25     Desiree Linden       Track    61.0      97
26   Shalane Flanagan       Track    65.0     106
27   Laurie Hernandez  Gymnastics    60.0     106

We see that Brittainey Raven's height is now the floating point value NaN meaning Not a Number. This NaN value is used even in columns that do not contain floating point values. For example, in row 20 above, Amy Cragg has NaN in the Sport column. Wes McKinney, the developer of Pandas calls NaN a *sentinel* value that is easily detected and indicates a missing value.

#### Adding a missing value by hand
Suppose we didn't know the number of cylinders of a Ford F150:

make | mpg | cylinders | HP | 0-60 |
---- | :---: | :---: | :---: | :---: 
Fiat | 38 | 4 | 157   | 6.9 
Ford F150 | 19 | - | 386 | 6.3 
Mazda 3 | 37 | 4 | 155 |  7.5 
Ford Escape | 27 | 4 | 245 | 7.1 
Kia Soul | 31 | 4 | 164 | 8.5 

In that case we can create a dataframe like:


In [ ]:
cars3 = {'make': Series(['Fiat 500', 'Ford F-150', 'Mazda 3', 'Ford Escape', 'Kia Soul']),
        'mpg': Series([38, 19, 37, 27, 31]),
        'cylinders': Series([4,np.nan, 4, 4, 4]),
        'HP': [157, 386, 155, 245, 164],
        '0-60': Series([6.9, 6.3, 7.5, 7.1, 8.5])}
        
carz = DataFrame(cars3)
carz

make  mpg  cylinders   HP  0-60
0     Fiat 500   38        4.0  157   6.9
1   Ford F-150   19        NaN  386   6.3
2      Mazda 3   37        4.0  155   7.5
3  Ford Escape   27        4.0  245   7.1
4     Kia Soul   31        4.0  164   8.5

where `np.nan` is Numpy's NaN.  We can also use Python's `None`:


In [ ]:
cars3 = {'make': Series(['Fiat 500', 'Ford F-150', 'Mazda 3', 'Ford Escape', 'Kia Soul']),
        'mpg': Series([38, 19, 37, 27, 31]),
        'cylinders': Series([4,None, 4, 4, 4]),
        'HP': [157, 386, 155, 245, 164],
        '0-60': Series([6.9, 6.3, 7.5, 7.1, 8.5])}
        
carz = DataFrame(cars3)
carz

make  mpg  cylinders   HP  0-60
0     Fiat 500   38        4.0  157   6.9
1   Ford F-150   19        NaN  386   6.3
2      Mazda 3   37        4.0  155   7.5
3  Ford Escape   27        4.0  245   7.1
4     Kia Soul   31        4.0  164   8.5

In addition to reading CSV files, there are many other ways of reading in data including from SQL databases, mongoDB, and webpages. See the Pandas documentation for details.

<h3 style="color:red">Q2. Pima Indians Diabetes Dataset</h3>
It is time to look at a new dataset, the Pima Indians Diabetes Data Set developed by the
United States National Institute of Diabetes and Digestive and Kidney Diseases.

The majority of the Pima people live on a reservation in Arizona.

![](https://upload.wikimedia.org/wikipedia/commons/6/65/Pinal_County_Arizona_Incorporated_and_Unincorporated_areas_GRIC_highlighted.svg)

Astonishingly, over 30% of Pima people develop diabetes. In contrast, the diabetes rate in
the United States is around 12% and in China it is 4.2%. The country with the lowest rate of diabetes is Benin at 1.5%.

Each instance in the dataset represents information about a Pima woman over the age of 21
and belonged to one of two classes: a person who developed diabetes within five years, or a
person that did not. There are eight attributes in addition to the column representing whether or not they developed diabetes:


1.  The number of times the woman was pregnant
2.  Plasma glucose concentration a 2 hours in an oral glucose tolerance test 
3.  Diastolic blood pressure (mm Hg)
4.  Triceps skin fold thickness (mm) 
5.  2-Hour serum insulin (mu U/ml) 
6.  Body mass index (weight in kg/(height in m)^2) 
7.  Diabetes pedigree function 
8.  Age
9.  Whether they got diabetes or not (0 = no, 1 = yes)


<span style="color:red">Please create a dataframe from the csv file at </span>

    https://raw.githubusercontent.com/zacharski/ml-class/master/data/pima-indians-diabetes.csv
    
<span style="color:red">This file does not have a header row</span>

In [ ]:
#Your code here
   

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   446  100   446    0     0    721      0 --:--:-- --:--:-- --:--:--   720


In [ ]:
## Your code here

## Accessing data in a DataFrame
We can get a column of a DataFrame by using the column name:

In [ ]:
athletes['Weight']

0      66
1     162
2     204
3      90
4      99
5     106
6     175
7     123
8      68
9     200
10    163
11     95
12     77
13    108
14    155
15    155
16    108
17    106
18     97
19     76
20     99
21    104
22    101
23    188
24    166
25     97
26    106
27    106
Name: Weight, dtype: int64

and we can get multiple columns by passing a list of column names

In [ ]:
athletes[['Weight', 'Height']]

Weight  Height
0       66      54
1      162      72
2      204      78
3       90      49
4       99      65
5      106      63
6      175      75
7      123      67
8       68      54
9      200      76
10     163      68
11      95      61
12      77      58
13     108      70
14     155      70
15     155      70
16     108      63
17     106      65
18      97      66
19      76      61
20      99      64
21     104      57
22     101      62
23     188      77
24     166      72
25      97      61
26     106      65
27     106      60

<h3 style="color:red">Q3. Pima Indians Diabetes Dataframe</h3>

<span style="color:red">Suppose we want to create a new dataframe `pima2` from the original one but with only the columns</span>



1.  Plasma glucose concentration a 2 hours in an oral glucose tolerance test 
2.  2-Hour serum insulin (mu U/ml) 
3.  Body mass index (weight in kg/(height in m)^2) 
4.  Diabetes pedigree function 
5.  Age
6.  Whether they got diabetes or not (0 = no, 1 = yes)


<span style="color:red">How would you do so?</span>


In [ ]:
# your code here

### Returning rows

To get a row we can use the `loc` function

In [ ]:
athletes.loc[0]

Name      Asuka Teramoto
Sport         Gymnastics
Height                54
Weight                66
Name: 0, dtype: object

We can also get rows that match a specific criterion. For example:

In [ ]:
basketballPlayers = athletes.loc[athletes['Sport'] == 'Basketball']
basketballPlayers

Name       Sport  Height  Weight
1    Brittainey Raven  Basketball      72     162
2            Chen Nan  Basketball      78     204
6       Jennifer Lacy  Basketball      75     175
9       Nakia Sanford  Basketball      76     200
10         Nikki Blue  Basketball      68     163
14    Shanna Crossley  Basketball      70     155
15   Shavonte Zellous  Basketball      70     155
23  Elena Delle Donne  Basketball      77     188
24   Seimone Augustus  Basketball      72     166

In [ ]:
tallBasketballPlayers = athletes.loc[(athletes['Sport'] == 'Basketball') & (athletes['Height'] > 72)]
tallBasketballPlayers

Name       Sport  Height  Weight
2            Chen Nan  Basketball      78     204
6       Jennifer Lacy  Basketball      75     175
9       Nakia Sanford  Basketball      76     200
23  Elena Delle Donne  Basketball      77     188

<span style="color:red">How would we create a new DataFrame that has all the athletes who weight under 100 pounds?</span>

Name       Sport  Height  Weight
0     Asuka Teramoto  Gymnastics      54      66
3      Gabby Douglas  Gymnastics      49      90
4   Helalia Johannes       Track      65      99
8        Linlin Deng  Gymnastics      54      68
11   Qiushuang Huang  Gymnastics      61      95
12    Rebecca Tunney  Gymnastics      58      77
18   Valeria Straneo       Track      66      97
19   Viktoria Komova  Gymnastics      61      76
20         Amy Cragg       Track      64      99
25    Desiree Linden       Track      61      97

We can add columns to a DataFrame.
For example the formula for Body Mass Index is

$$BMI = \frac{weightInPounds}{heightInInches^2} \times 703 $$

In [ ]:
athletes['bmi'] = 703 *athletes['Weight'] / (athletes['Height']**2) 

In [ ]:
athletes

Name       Sport  Height  Weight        bmi
0      Asuka Teramoto  Gymnastics      54      66  15.911523
1    Brittainey Raven  Basketball      72     162  21.968750
2            Chen Nan  Basketball      78     204  23.571992
3       Gabby Douglas  Gymnastics      49      90  26.351520
4    Helalia Johannes       Track      65      99  16.472663
5     Irina Miketenko       Track      63     106  18.775006
6       Jennifer Lacy  Basketball      75     175  21.871111
7        Kara Goucher       Track      67     123  19.262419
8         Linlin Deng  Gymnastics      54      68  16.393690
9       Nakia Sanford  Basketball      76     200  24.342105
10         Nikki Blue  Basketball      68     163  24.781358
11    Qiushuang Huang  Gymnastics      61      95  17.948132
12     Rebecca Tunney  Gymnastics      58      77  16.091260
13        Rene Kalmer       Track      70     108  15.494694
14    Shanna Crossley  Basketball      70     155  22.237755
15   Shavonte Zellous  Basketball      70     155  22.237755
16    Tatyana Petrova       Track      63     108  19.129252
17        Tiki Gelana       Track      65     106  17.637396
18    Valeria Straneo       Track      66      97  15.654500
19    Viktoria Komova  Gymnastics      61      76  14.358506
20          Amy Cragg       Track      64      99  16.991455
21       Simone Biles  Gymnastics      57     104  22.502924
22     Madison Kocian  Gymnastics      62     101  18.471124
23  Elena Delle Donne  Basketball      77     188  22.291111
24   Seimone Augustus  Basketball      72     166  22.511188
25     Desiree Linden       Track      61      97  18.325988
26   Shalane Flanagan       Track      65     106  17.637396
27   Laurie Hernandez  Gymnastics      60     106  20.699444

## Descriptive Statistics on DataFrames
One handy function is `describe`

In [ ]:
athletes.describe()

Height      Weight        bmi
count  28.000000   28.000000  28.000000
mean   65.107143  121.428571  19.640072
std     7.258912   40.274455   3.264325
min    49.000000   66.000000  14.358506
25%    61.000000   97.000000  16.861757
50%    65.000000  106.000000  18.952129
75%    70.000000  156.750000  22.251094
max    78.000000  204.000000  26.351520

Alternatively, I could ask for a specific statistic:


In [ ]:
athletes['Weight'].mean()

121.42857142857143

In [ ]:
athletes.loc[athletes['Sport'] == 'Gymnastics' ]['Weight'].mean()

87.0

In [ ]:
athletes.loc[athletes['Sport'] == 'Basketball' ]['Weight'].mean()

174.22222222222223

<h3 style="color:red">Q4. Pima Indians</h3>
<span style="color:red">I would like to fill out this little table:</span>

  x | Avg. BMI | Avg. Diabetes Pedigree | Avg. times pregnant | Avg. Plasma glucose |
---   | :---: | :---: | :---: | :---: |
Has Diabetes |   | |  |
Doesn't have Diabetes |   | |  |

<span style="color:red">Can you get this information (where 'Avg' refers to the mean)? So, for example, the first cell is the average Body Mass Index of people who have diabetes.</span> You don't need to write code to generate a table, I just want to see the values that would be put in such a table.


In [ ]:
# your code here

### Music Ratings

Suppose I have customers of my vinyl record shop rate different artists

|Customer | Taylor Swift | Miranda Lambert | Carrie Underwood | Nicki Minaj | Ariana Grande |
|:-----------|:------:|:------:|:---------:|:------:|:--------:|
|Jake|5|-|5|2|2|
|Clara|2|-|-|4|5|
|Kelsey|5|5|5|2|-|
|Angelica|2|3|-|5|5|
|Jordyn|2|1|-|5|-|

<h3 style="color:red">Q4. Ratings</h3>
<span style="color:red">Create a dataframe called `ratings` from the CSV file</span>

    https://raw.githubusercontent.com/zacharski/ml-class/master/data/ratings.csv

In [ ]:
# your code here
ratings = "to do"

Customer  Taylor Swift  Miranda Lambert  Carrie Underwood  Nicki Minaj  \
0      Jake             5              NaN               5.0            2   
1     Clara             2              NaN               NaN            4   
2    Kelsey             5              5.0               5.0            2   
3  Angelica             2              3.0               NaN            5   
4    Jordyn             2              1.0               NaN            5   

   Ariana Grande  
0            2.0  
1            5.0  
2            NaN  
3            5.0  
4            NaN

We can get the mean rating of each artist by:


In [ ]:
ratings.mean()

Taylor Swift        3.2
Miranda Lambert     3.0
Carrie Underwood    5.0
Nicki Minaj         3.6
Ariana Grande       4.0
dtype: float64

Many descriptive statistics functions take an optional parameter `axis` that tells which axis to reduce over. If we want the mean ratings for each **customer** instead of each artist we can do:

In [ ]:
ratings.mean(axis=1)

0    3.500000
1    3.666667
2    4.250000
3    3.750000
4    2.666667
dtype: float64

Well, that was sort of unhelpful. We know that person 2 has a mean of 4.25, but it would be nice to use names as the indices for the rows instead of numbers. So right now we have the rows labeled 0, 1, 2, 3 ... but we would like them labeled *Jake, Clara, Kelsey*   We can do so by creating a new dataframe:

In [ ]:
ratings2 = ratings.set_index('Customer')
ratings2.mean(axis=1)

Customer
Jake        3.500000
Clara       3.666667
Kelsey      4.250000
Angelica    3.750000
Jordyn      2.666667
dtype: float64

*Note that while this looks like it erroneously creates an additional row called Customer, that is just the way the table is displayed*

Sweet! So `axis=1` means reduce by rows and `axis=0` means reduce by columns:

In [ ]:
ratings2.mean(axis=0)

Taylor Swift        3.2
Miranda Lambert     3.0
Carrie Underwood    5.0
Nicki Minaj         3.6
Ariana Grande       4.0
dtype: float64

## list of descriptive statistics
(from the book *Python for Data Analysis*)

Method | Description
:-- | :--
`count` | Number of non-NaN values
`describe` | A set of common summary statistics
`min, max` | compute minimum and maximum values
`argmin, argmax` | compute index locations of minimum and maximum values
`sum` | Sum the values
`mean` | Mean of values
`median` | Median of values
`std` | Sample standard deviation

So, for example, the lowest rating for each artist:


In [ ]:
ratings2.min()

Taylor Swift        2.0
Miranda Lambert     1.0
Carrie Underwood    5.0
Nicki Minaj         2.0
Ariana Grande       2.0
dtype: float64

<h3 style="color:red">Q5. Ratings 2</h3>
<span style="color:red">What are the median ratings for each customer?</span>

In [ ]:
# your code

## sorting
To sort by the index we can use the `sort_index` method

In [ ]:
athlete = pd.read_csv('https://raw.githubusercontent.com/zacharski/ml-class/master/data/athletesNoHeader.csv', names=['Name', 'Sport', 'Height', 'Weight'])
a = athlete.set_index('Name')
a.sort_index()

Sport  Height  Weight
Name                                         
Amy Cragg               Track      64      99
Asuka Teramoto     Gymnastics      54      66
Brittainey Raven   Basketball      72     162
Chen Nan           Basketball      78     204
Desiree Linden          Track      61      97
Elena Delle Donne  Basketball      77     188
Gabby Douglas      Gymnastics      49      90
Helalia Johannes        Track      65      99
Irina Miketenko         Track      63     106
Jennifer Lacy      Basketball      75     175
Kara Goucher            Track      67     123
Laurie Hernandez   Gymnastics      60     106
Linlin Deng        Gymnastics      54      68
Madison Kocian     Gymnastics      62     101
Nakia Sanford      Basketball      76     200
Nikki Blue         Basketball      68     163
Qiushuang Huang    Gymnastics      61      95
Rebecca Tunney     Gymnastics      58      77
Rene Kalmer             Track      70     108
Seimone Augustus   Basketball      72     166
Shalane Flanagan        Track      65     106
Shanna Crossley    Basketball      70     155
Shavonte Zellous   Basketball      70     155
Simone Biles       Gymnastics      57     104
Tatyana Petrova         Track      63     108
Tiki Gelana             Track      65     106
Valeria Straneo         Track      66      97
Viktoria Komova    Gymnastics      61      76

We can sort in reverse by

In [ ]:
a.sort_index(ascending=False)

Sport  Height  Weight
Name                                         
Viktoria Komova    Gymnastics      61      76
Valeria Straneo         Track      66      97
Tiki Gelana             Track      65     106
Tatyana Petrova         Track      63     108
Simone Biles       Gymnastics      57     104
Shavonte Zellous   Basketball      70     155
Shanna Crossley    Basketball      70     155
Shalane Flanagan        Track      65     106
Seimone Augustus   Basketball      72     166
Rene Kalmer             Track      70     108
Rebecca Tunney     Gymnastics      58      77
Qiushuang Huang    Gymnastics      61      95
Nikki Blue         Basketball      68     163
Nakia Sanford      Basketball      76     200
Madison Kocian     Gymnastics      62     101
Linlin Deng        Gymnastics      54      68
Laurie Hernandez   Gymnastics      60     106
Kara Goucher            Track      67     123
Jennifer Lacy      Basketball      75     175
Irina Miketenko         Track      63     106
Helalia Johannes        Track      65      99
Gabby Douglas      Gymnastics      49      90
Elena Delle Donne  Basketball      77     188
Desiree Linden          Track      61      97
Chen Nan           Basketball      78     204
Brittainey Raven   Basketball      72     162
Asuka Teramoto     Gymnastics      54      66
Amy Cragg               Track      64      99

We can use the `sort_values` method with the `by` argument to sort by a particular column:

In [ ]:
a.sort_values(by='Sport')

Sport  Height  Weight
Name                                         
Brittainey Raven   Basketball      72     162
Chen Nan           Basketball      78     204
Seimone Augustus   Basketball      72     166
Elena Delle Donne  Basketball      77     188
Jennifer Lacy      Basketball      75     175
Nakia Sanford      Basketball      76     200
Nikki Blue         Basketball      68     163
Shavonte Zellous   Basketball      70     155
Shanna Crossley    Basketball      70     155
Asuka Teramoto     Gymnastics      54      66
Madison Kocian     Gymnastics      62     101
Simone Biles       Gymnastics      57     104
Viktoria Komova    Gymnastics      61      76
Laurie Hernandez   Gymnastics      60     106
Rebecca Tunney     Gymnastics      58      77
Qiushuang Huang    Gymnastics      61      95
Linlin Deng        Gymnastics      54      68
Gabby Douglas      Gymnastics      49      90
Tatyana Petrova         Track      63     108
Tiki Gelana             Track      65     106
Valeria Straneo         Track      66      97
Amy Cragg               Track      64      99
Kara Goucher            Track      67     123
Irina Miketenko         Track      63     106
Helalia Johannes        Track      65      99
Desiree Linden          Track      61      97
Shalane Flanagan        Track      65     106
Rene Kalmer             Track      70     108

In [ ]:
a.sort_values(by=['Height', 'Weight'])

Sport  Height  Weight
Name                                         
Gabby Douglas      Gymnastics      49      90
Asuka Teramoto     Gymnastics      54      66
Linlin Deng        Gymnastics      54      68
Simone Biles       Gymnastics      57     104
Rebecca Tunney     Gymnastics      58      77
Laurie Hernandez   Gymnastics      60     106
Viktoria Komova    Gymnastics      61      76
Qiushuang Huang    Gymnastics      61      95
Desiree Linden          Track      61      97
Madison Kocian     Gymnastics      62     101
Irina Miketenko         Track      63     106
Tatyana Petrova         Track      63     108
Amy Cragg               Track      64      99
Helalia Johannes        Track      65      99
Tiki Gelana             Track      65     106
Shalane Flanagan        Track      65     106
Valeria Straneo         Track      66      97
Kara Goucher            Track      67     123
Nikki Blue         Basketball      68     163
Rene Kalmer             Track      70     108
Shanna Crossley    Basketball      70     155
Shavonte Zellous   Basketball      70     155
Brittainey Raven   Basketball      72     162
Seimone Augustus   Basketball      72     166
Jennifer Lacy      Basketball      75     175
Nakia Sanford      Basketball      76     200
Elena Delle Donne  Basketball      77     188
Chen Nan           Basketball      78     204

In that last example, we first sorted by height and then by weight.

# The nearest neighbor Music example in Pandas

Let's go back to our Numpy task that dealt with the following music data:



| Guest  | Janelle Monae  | Major Lazer  | Tim McGraw  |  Maren Morris | Hailee Steinfeld| 
|---|---|---|---|---|---|
|  Ann | 4  |  5 | 2  |  1 | 3 |
| Ben  |  3 |  1 |  5 | 4  | 2|
| Jordyn  | 5  |  5 | 2  | 2  | 3|
|  Sam | 4 | 1 | 4 | 4 | 1|
| Hyunseo | 1 | 1 | 5 | 4 | 1 |
| Lauren | 3 | 1 | 5 |  5 | 5 |
|  Ahmed | 4  |  5 | 3  |  3 | 1 |


I want to find out who is the most similar to Mikaela and who is most similar to Brandon, who rated the artists:


| Guest  | Janelle Monae  | Major Lazer  | Tim McGraw  |  Maren Morris | Hailee Steinfeld| 
|---|---|---|---|---|---|
|  Mikaela | 3  |  2 | 4  |  5 |4 |
|  Brandon | 4  |  5 | 1  |  2 |3 |

My first step is to read in the data and convert the guest name column as the index:

In [ ]:
f = pd.read_csv('https://raw.githubusercontent.com/zacharski/ml-class/master/data/ratings2.csv')
fr = f.set_index('Guest')
fr

Janelle Monae  Major Lazer  Tim McGraw  Maren Morris  \
Guest                                                           
Ann                  4            5           2             1   
Ben                  3            1           5             4   
Jordyn               5            5           1             1   
Sam                  4            1           4             4   
Hyunseo              1            1           5             4   
Lauren               3            1           5             5   
Ahmed                4            5           3             3   

         Hailee Steinfeld  
Guest                      
Ann                     3  
Ben                     2  
Jordyn                  3  
Sam                     1  
Hyunseo                 1  
Lauren                  5  
Ahmed                   1

Now let's load in Mikaela and Brandon in a different DataFrame

In [ ]:
mb = pd.read_csv('https://raw.githubusercontent.com/zacharski/ml-class/master/data/ratings3.csv')
mb.set_index('Guest', inplace=True)
mb

Janelle Monae  Major Lazer  Tim McGraw  Maren Morris  \
Guest                                                           
Mikaela              3            2           4             5   
Brandon              4            5           1             2   

         Hailee Steinfeld  
Guest                      
Mikaela                 4  
Brandon                 3

Let's work through the example step-by-step.

First, can you set a variable named `mikaela` to equal the Mikaela row of `mb`?
So something that starts with

    mikaela = mb???????????

First, let's subtract Mikaela's ratings from the other people's:

In [ ]:
fr2 =fr.sub(mikaela)
fr2

Janelle Monae  Major Lazer  Tim McGraw  Maren Morris  \
Guest                                                           
Ann                  1            3          -2            -4   
Ben                  0           -1           1            -1   
Jordyn               2            3          -3            -4   
Sam                  1           -1           0            -1   
Hyunseo             -2           -1           1            -1   
Lauren               0           -1           1             0   
Ahmed                1            3          -1            -2   

         Hailee Steinfeld  
Guest                      
Ann                    -1  
Ben                    -2  
Jordyn                 -1  
Sam                    -3  
Hyunseo                -3  
Lauren                  1  
Ahmed                  -3



Next, we will get the absolute values 

In [ ]:
fr3 = fr2.apply(np.abs)
fr3

Janelle Monae  Major Lazer  Tim McGraw  Maren Morris  \
Guest                                                           
Ann                  1            3           2             4   
Ben                  0            1           1             1   
Jordyn               2            3           3             4   
Sam                  1            1           0             1   
Hyunseo              2            1           1             1   
Lauren               0            1           1             0   
Ahmed                1            3           1             2   

         Hailee Steinfeld  
Guest                      
Ann                     1  
Ben                     2  
Jordyn                  1  
Sam                     3  
Hyunseo                 3  
Lauren                  1  
Ahmed                   3

and sum the values in each row to compute the distances

In [ ]:
fr4 = fr3.sum(axis = 1)
fr4

Guest
Ann        11
Ben         5
Jordyn     13
Sam         6
Hyunseo     8
Lauren      3
Ahmed      10
dtype: int64

Finally, let's sort by the distance and get the three closest guests:

In [ ]:
fr4.sort_values().head(3)

Guest
Lauren    3
Ben       5
Sam       6
dtype: int64

So Lauren is the person most similar to Mikaela, followed by Ben and Sam.

<h3 style="color:red">Q6. Find Closest</h3>

<span style="color:red">Can you write a function nearestNeighbor that takes 2 arguments: a customers dataframe,  and an array representing one customers ratings and returns the name and distance of the closest  customer (using Euclidean Distance)?</span>



In [ ]:
ahmed = [4, 5, 3, 3, 1]
def nearestNeighbor(ratings, customer):
    # Your code here
    return 'result'
    
    
print(nearestNeighbor(fr, mikaela))
# TO DO
# brandon = ...
print(nearestNeighbor(fr, mb.loc['Brandon']))

Guest
Lauren    3
dtype: int64
Guest
Ann    2
dtype: int64


<h3 style="color:red">Q7. Find Closest using Euclidean Distance</h3>

<span style="color:red">Can you write a new function nearestNeighbor that takes 3 arguments</span>

* ratings - an DataFrame representing all our customer ratings
* customer - a DataFrame representing one customer's rating
* metric - the value is either the string 'manhattan' or 'euclidean' representing which metric to use to compute the distance.
    a customers dataframe,  and an array representing one customers ratings
    
It returns the name and distance of the closest  customer (using Euclidean Distance)?


In [ ]:
print(nearestNeighbor(fr, mikaela, metric='euclidean'))
# TO DO
# brandon = ...
print(nearestNeighbor(fr, mb.loc['Brandon'], metric='euclidean'))